api/v1beta1/openstackdataplanenodeset_types.go

```diff
@@ -204,3 +209,28 @@ type DataplaneAnsibleImageDefaults struct {
 	TelemetryCeilometerIpmi    string
 	TelemetryNodeExporter      string
 }
+
+// duplicateNodeCheck checks the NodeSetList for pre-existing nodes. If the user is trying to redefine an
+// existing node, we will return an error and block resource creation.
+func (r *OpenStackDataPlaneNodeSetSpec) duplicateNodeCheck(nodeSetList *OpenStackDataPlaneNodeSetList) (errors field.ErrorList) {
+	existingNodeNames := make([]string, 0)
+	for _, existingNode := range nodeSetList.Items {
+		for _, node := range existingNode.Spec.Nodes {
+			existingNodeNames = append(existingNodeNames, node.HostName)
+			if node.Ansible.AnsibleHost != "" {
+				existingNodeNames = append(existingNodeNames, node.Ansible.AnsibleHost)
+			}
+		}
+	}
+
+	for _, newNodeName := range r.Nodes {
+		if slices.Contains(existingNodeNames, newNodeName.HostName) || slices.Contains(existingNodeNames, newNodeName.Ansible.AnsibleHost) {
+			errors = append(errors, field.Invalid(
+				field.NewPath("Spec").Child("nodes"),
+				newNodeName,
+				fmt.Sprintf("node already exists in the cluster: %s", newNodeName.HostName)))
```

Comment:
As a future improvement I would tell the user which NodeSet contains the definition of this node.

Reply:
Yes, I would leave that for the future.